### PR TITLE
Bug 1784264: add namespace level imagestream fetching

### PR DESCRIPTION
**Pull request by karthikjeeyar**
User with limited permissions (few namespaces) will be able to pull the images from namespaces accordingly.

user with limited permissions to namespaces:
test · Details · OKD
`

AwesomeScreenshot-2019-12-26-1577335013068

Fixes: https://issues.redhat.com/browse/ODC-2536

### CODE REVIEW

File: frontend/packages/dev-console/src/components/import/image-search/ImageStreamNsDropdown.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  return (
+    <ResourceDropdownField
+      name="imageStream.namespace"
+      label="Projects"
+      title="Select Project"
+      fullWidth
+      required
+      resources={getProjectResource()}
+      dataSelector={['metadata', 'name']}
+      onChange={(selectedProject) => {
```
Comment (vikram-raj) [inline frontend/packages/dev-console/src/components/import/image-search/ImageStreamNsDropdown.tsx]: Can you make it a function and use it here.

File: frontend/packages/dev-console/src/components/import/import-types.ts [hunk 2]
```diff
@@ -9,10 +9,32 @@ export interface DeployImageFormProps {
...
 }
 
-export interface ImageStreamProps {
-  imageStreams: K8sResourceKind[];
+export type ImageStreamState = {
+  hasAccessToPullImage: boolean;
+  accessLoading: boolean;
+  loading: boolean;
+  hasCreateAccess: boolean;
+  selectedImageStream: K8sResourceKind | any;
```
Comment (vikram-raj) [inline frontend/packages/dev-console/src/components/import/import-types.ts]: Do we need type ```any``` here. Isn't ```selectedImageStream``` is always of type ```K8sResourceKind```.

File: frontend/packages/dev-console/src/components/import/import-types.ts [hunk 3]
```diff
@@ -9,10 +9,32 @@ export interface DeployImageFormProps {
...
+  setAccessLoading = 'setAccessLoading',
+  setLoading = 'setLoading',
+  setSelectedImageStream = 'setSelectedImageStream',
+  setHasAccessToPullImage = 'setHasAccessToPullImage',
+  setHasCreateAccess = 'setHasCreateAccess',
+}
+export type ImageStreamAction =
+  | { type: ImageStreamActions.setAccessLoading; value: boolean }
+  | { type: ImageStreamActions.setLoading; value: boolean }
+  | { type: ImageStreamActions.setSelectedImageStream; value: K8sResourceKind | any }
```
Comment (vikram-raj) [inline frontend/packages/dev-console/src/components/import/import-types.ts]: Isn't ```value``` is always of kind ```K8sResourceKind```.

Comment (karthikjeeyar) [reply frontend/packages/dev-console/src/components/import/import-types.ts]: Yes It is, but it throws weird  lint error since we are using `|` statement. The first `setAccessLoading` statement is also picking up the `K8sResourceKind` | `boolean`, thats why i add any here to avoid lint error. but i agree, this needs to be fixed.

Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/import/import-types.ts]: Can we simply define the type like this 
```export type ImageStreamAction = { type: ImageStreamActions; value: boolean | K8sResourceKind };```

File: frontend/packages/dev-console/src/components/import/import-types.ts (hunk 2 above)
Comment (karthikjeeyar) [reply frontend/packages/dev-console/src/components/import/import-types.ts]: Again, this is added due to the lint error which happened for union of types. I fixed it along with the below comment.

File: frontend/public/components/utils/webhooks.tsx [hunk 4]
```diff
@@ -90,7 +90,7 @@ export const WebhookTriggers: React.FC<WebhookTriggersProps> = (props) => {
       setWebhookSecrets(_.compact(secrets));
       setLoaded(true);
     });
-  }, [isLoaded, canGetSecret]);
+  }, [isLoaded, canGetSecret, secretNames, namespace]);
```
Comment (vikram-raj) [inline frontend/public/components/utils/webhooks.tsx]: Undo this. This change is not related to this PR.

Comment (karthikjeeyar) [reply frontend/public/components/utils/webhooks.tsx]: Ahh, yarn lint --fix must have added it.

**Comment by karthikjeeyar** [comment]
/retitle Bug 1784264: add namespace level imagestream fetching

File: frontend/packages/dev-console/src/components/import/image-search/ImageStreamNsDropdown.tsx [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            name: 'system:image-puller',
+            namespace: selectedProject,
+          })
+            .then((resp) =>
+              dispatch({ type: Action.setHasCreateAccess, value: resp.status.allowed }),
+            )
+            .catch(() => dispatch({ type: Action.setHasAccessToPullImage, value: false })),
+        );
+        promiseArr.push(
+          k8sGet(RoleBindingModel, 'system:image-puller', selectedProject)
```
Comment (vikram-raj) [inline frontend/packages/dev-console/src/components/import/image-search/ImageStreamNsDropdown.tsx]: Getting 404 for this ```system:image-puller```. 

Screenshot from 2020-01-03 14-07-32

Comment (karthikjeeyar) [reply frontend/packages/dev-console/src/components/import/image-search/ImageStreamNsDropdown.tsx]: This api call to figure out whether we have access to grant permission, if the api throws 404 then catch block will set the grantAccess variable to false.

Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/import/image-search/ImageStreamNsDropdown.tsx]: Got it, Thanks.

File: frontend/packages/console-shared/src/components/formik-fields/ResourceDropdownField.tsx [hunk 6]
```diff
@@ -10,6 +10,8 @@ import { getFieldId } from './field-utils';
 export interface ResourceDropdownFieldProps extends DropdownFieldProps {
   dataSelector: string[] | number[] | symbol[];
   resources: FirehoseResource[];
+  onLoad?: (items: { [key: string]: string }) => void;
+  resourceFilter?: (resource: any) => boolean;
```
Comment (invincibleJai) [inline frontend/packages/console-shared/src/components/formik-fields/ResourceDropdownField.tsx]: nit : we can have type here

**Comment by invincibleJai** [comment]
/lgtm

Verified locally seems to works as expected.

File: frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx [hunk 7]
```diff
@@ -2,79 +2,67 @@ import * as React from 'react';
...
+    case ImageStreamActions.setLoading:
+      return { ...state, loading: value };
+    case ImageStreamActions.setAccessLoading:
+      return { ...state, accessLoading: value };
+    case ImageStreamActions.setHasCreateAccess:
+      return { ...state, hasCreateAccess: value };
+    case ImageStreamActions.setSelectedImageStream:
+      return { ...state, selectedImageStream: value };
+    default:
+      return initialState;
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx]: This is a closed React reducer, perhaps throw an error here so there is no accidental mis-usage. 

Got the idea from here: https://reactjs.org/docs/hooks-reference.html#usereducer

File: frontend/packages/dev-console/src/utils/imagestream-utils.ts [hunk 8]
```diff
@@ -212,3 +191,16 @@ export const getProjectResource = (): FirehoseResource[] => {
     },
   ];
 };
+
+export const getImageStreamResource = (namespace: string): FirehoseResource[] => {
+  const resource = [];
+  if (namespace && namespace !== '') {
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/utils/imagestream-utils.ts]: `namespace` cannot be truthy while `namespace` is equal to an empty string... this is a duplicate check.
```suggestion
  if (namespace) {
```

File: frontend/packages/dev-console/src/utils/imagestream-utils.ts [hunk 9]
```diff
@@ -174,35 +174,14 @@ export const getSortedTags = (imageStream: K8sResourceKind) => {
       })
     : [];
 };
-export const getImageStreamTags = (imageStreams: K8sResourceKind[], name: string, ns: string) => {
-  const imageStream = _.find(imageStreams, (img) => {
-    return img.metadata.name === name && img.metadata.namespace === ns;
-  });
-  const sortedTags =
-    name && imageStream && !_.isEmpty(imageStream) ? getSortedTags(imageStream) : [];
+export const getImageStreamTags = (imageStream: K8sResourceKind | {}) => {
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/utils/imagestream-utils.ts]: Can `imageStream` be anything but a `K8sResourceKind`?
```suggestion
export const getImageStreamTags = (imageStream: K8sResourceKind) => {
```
Is there a specific reason you want to allow an empty object as a valid type?

Comment (karthikjeeyar) [reply frontend/packages/dev-console/src/utils/imagestream-utils.ts]: No Imagestream should be an K8sResourceKind only. I have removed this and now type casting the argument to `K8sResourceKind` from the caller function. Thanks Andrew i missed to change it.

**Comment by andrewballantyne** [comment]
At least the frontend pass locally... not sure what is up.

/retest

**Comment by andrewballantyne** [comment]
Ah, there is a TS error:

> `/frontend/packages/dev-console/src/components/import/image-search/ImageStreamTagDropdown.tsx`(24,43):
TS2345: Argument of type 'ImageStreamPayload' is not assignable to parameter of type 'K8sResourceKind'.
  Type 'false' has no properties in common with type 'K8sResourceKind'.

**Comment by karthikjeeyar** [comment]
> Ah, there is a TS error:
> 
> > `/frontend/packages/dev-console/src/components/import/image-search/ImageStreamTagDropdown.tsx`(24,43):
> > TS2345: Argument of type 'ImageStreamPayload' is not assignable to parameter of type 'K8sResourceKind'.
> > Type 'false' has no properties in common with type 'K8sResourceKind'.

yarn lint is not catching this error, fixed it now
cc: @andrewballantyne

**Comment by andrewballantyne** [comment]
@karthikjeeyar there are 3 points, TypeScript (`yarn dev`), Lint (`yarn lint`) and Unit Tests (`yarn test`) that incorporate the "frontend" tests.
